src/main/java/com/mentorship/vineservice/repositories/impl/VineRepositoryImpl.java

```diff
@@ -0,0 +1,68 @@
+package com.mentorship.vineservice.repositories.impl;
+
+import com.mentorship.vineservice.domain.Vine;
+import com.mentorship.vineservice.dto.PaginationDto;
+import com.mentorship.vineservice.repositories.VineRepositoryCustom;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.PersistenceContext;
+import jakarta.persistence.TypedQuery;
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Predicate;
+import jakarta.persistence.criteria.Root;
+import java.util.ArrayList;
+import java.util.List;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class VineRepositoryImpl implements VineRepositoryCustom {
+
+    @PersistenceContext
+    private EntityManager entityManager;
+
+
+    @Override
+    public Page<Vine> findVinesBySugarAndColorAndTypeAndGrapePaginated(PaginationDto pagination, String sugar,
+        String color, String name, String grape) {
```

Comment:
Did you think about implementation such a logic with JpaSpecification. I think the structure of your code would look a little bit more concise and understandable